File: ios/TimberCruisingApp/InventoryEngine/ExpansionFactors.swift
// Spec §7.6 + §1 Glossary.
//
// Expansion factor (EF):
//   - fixed-area:     EF = 1 / plot_area_acres  (trees per acre per tree)
//   - variable-radius (BAF): per-tree EF = BAF / BA_tree   (trees per acre
//     per "in" tree).
//
// These functions are shared between the live plot-tally display and the
// stand aggregation pipeline.

import Foundation
import Models

public enum ExpansionFactors {

    /// §7.6 Fixed-area plot expansion factor: trees/acre per tallied tree.
    /// Input: plot area in acres. Output: expansion factor.
    public static func fixedArea(plotAreaAcres: Float) -> Float {
        precondition(plotAreaAcres > 0, "plotAreaAcres must be > 0")
        return 1 / plotAreaAcres
    }

    /// §7.6 BAF expansion factor for a single tree: trees/acre per "in" tree.
    /// EF_i = BAF / BA_i.
    public static func variableRadius(baf: Float, dbhCm: Float) -> Float {
        precondition(baf > 0, "baf must be > 0")
        let ba = basalAreaM2(dbhCm: dbhCm)
        precondition(ba > 0, "DBH must be > 0")
        return baf / ba
    }

    /// Convenience: per-acre attribute total for a fixed-area plot.
    /// `attr(tree)` returns the per-tree value (e.g., volume, BA, count=1).
    public static func perAcreFixedArea<T>(
        trees: [Tree],
        plotAreaAcres: Float,
        attr: (Tree) -> T
    ) -> T where T: AdditiveArithmetic {
        let live = trees.filter { $0.deletedAt == nil }
        let sum = live.reduce(T.zero) { $0 + attr($1) }
        // scale by EF (constant across all trees in fixed-area plot)
        return scale(sum, by: fixedArea(plotAreaAcres: plotAreaAcres))
    }

    /// Convenience: per-acre attribute total for a BAF plot (each tree has
    /// its own EF).
    public static func perAcreBAF(
        trees: [Tree],
        baf: Float,
        attr: (Tree) -> Float
    ) -> Float {
        let live = trees.filter { $0.deletedAt == nil }
        return live.reduce(0) { $0 + attr($1) * variableRadius(baf: baf, dbhCm: $1.dbhCm) }
    }

    // Float-only scale helper because AdditiveArithmetic has no multiplication.
    private static func scale<T>(_ value: T, by factor: Float) -> T where T: AdditiveArithmetic {
        if let v = value as? Float, let r = (v * factor) as? T { return r }
        if let v = value as? Double, let r = (v * Double(factor)) as? T { return r }
        if let v = value as? Int {
            // Rare: callers should pass Float/Double; keep behavior predictable.
            if let r = Int(Float(v) * factor) as? T { return r }
        }
        return value
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/inventory/ExpansionFactors.kt
// Port of iOS InventoryEngine/ExpansionFactors.swift.
// Spec §7.6 + §1 Glossary.
//
// Expansion factor (EF):
//   - fixed-area:     EF = 1 / plot_area_acres  (trees per acre per tree)
//   - variable-radius (BAF): per-tree EF = BAF / BA_tree   (trees per acre
//     per "in" tree), with BAF and BA_tree BOTH in square feet.
//
// These functions are shared between the live plot-tally display and the
// stand aggregation pipeline.

package com.hcjeong.forestix.inventory

import com.hcjeong.forestix.data.cruise.Tree

object ExpansionFactors {

    /// §7.6 Fixed-area plot expansion factor: trees/acre per tallied tree.
    /// Input: plot area in acres. Output: expansion factor.
    fun fixedArea(plotAreaAcres: Float): Float {
        require(plotAreaAcres > 0f) { "plotAreaAcres must be > 0" }
        return 1f / plotAreaAcres
    }

    /// §7.6 BAF expansion factor for a single tree: trees/acre per "in" tree.
    /// EF_i = BAF / BA_i.
    ///
    /// BOTH SIDES IN SQUARE FEET. `baf` is ft²/ac — see the unit note at the
    /// top of `BasalAreaMath.kt` — so the tree's area comes from
    /// `basalAreaFt2`, not the metric `basalAreaM2` this used to divide by.
    /// The ratio of the two is 10.76, and it lands on every TPA, basal area
    /// and volume a prism plot reports.
    fun variableRadius(baf: Float, dbhCm: Float): Float {
        require(baf > 0f) { "baf must be > 0" }
        val ba = basalAreaFt2(dbhCm = dbhCm)
        require(ba > 0f) { "DBH must be > 0" }
        return baf / ba
    }

    /// Convenience: per-acre attribute total for a fixed-area plot.
    /// `attr(tree)` returns the per-tree value (e.g., volume, BA, count=1).
    /// (Swift is generic over AdditiveArithmetic; specialized to Float here —
    /// every call site uses Float.)
    fun perAcreFixedArea(
        trees: List<Tree>,
        plotAreaAcres: Float,
        attr: (Tree) -> Float,
    ): Float {
        val live = trees.filter { it.deletedAt == null }
        val sum = live.fold(0f) { acc, t -> acc + attr(t) }
        // scale by EF (constant across all trees in fixed-area plot)
        return sum * fixedArea(plotAreaAcres = plotAreaAcres)
    }

    /// Convenience: per-acre attribute total for a BAF plot (each tree has
    /// its own EF).
    fun perAcreBAF(
        trees: List<Tree>,
        baf: Float,
        attr: (Tree) -> Float,
    ): Float {
        val live = trees.filter { it.deletedAt == null }
        return live.fold(0f) { acc, t ->
            acc + attr(t) * variableRadius(baf = baf, dbhCm = t.dbhCm)
        }
    }
}
